File: sources/faf/game_prefs_to_hotkey_json.py

import re
import json
from datetime import datetime


def extract_hotkeys_from_lua(lua_content):
    user_key_map_pattern = r'UserKeyMap\s*=\s*{(.*?)}'
    user_key_map_content = re.search(user_key_map_pattern, lua_content, re.DOTALL).group(1)
    pattern = r'(\[[\'"]?[\w\-\+\s]+[\'"]?\]|[\w\-]+)\s*=\s*\'([^\']+)\''
    return re.findall(pattern, user_key_map_content)


lua_to_json_format = {
    "[": "",
    "]": "",
    "'": "",
    "\"": "",
    "-": " + ",
    "LeftBracket": "[",
    "RightBracket": "]",
    "Equals": "=",
    "NumStar": "*",
    "Quote": "'",
    "NumSlash": "/",
    "Slash": "/",
    "Tilde": "`",
    "Backslash": "\\",
    "Semicolon": ";",
    "Period": "."
}


def convert_lua_to_json_format(key, format):
    for lua_key_part, json_key_part in format.items():
        key = key.replace(lua_key_part, json_key_part)
    return key


def replace_all(text, dic):
    for i, j in dic.items():
        text = text.replace(i, j)
    return text


def main(game_prefs_filepath="./CheeseBerry_Game.prefs"):
    year_month_day = datetime.now().strftime('%Y-%m-%d')
    keymap_name = replace_all(
        game_prefs_filepath,
        {"_Game": "",
         ".prefs": "",
         ".": "",
         "/": "",
         "\\": ""
         }
    )
    output_filename = keymap_name + "_" + year_month_day + ".json"

    # Extract hotkeys from the .prefs file and print them
    with open(game_prefs_filepath, 'r') as f:
        prefs_content = f.read()
    extracted_hotkeys = extract_hotkeys_from_lua(prefs_content)

    # Translate lua to json hotkey format and print the updated list
    updated_hotkeys = {}
    for key, action in extracted_hotkeys:
        key = convert_lua_to_json_format(key, lua_to_json_format)
        updated_hotkeys[action] = [key, ""]

    # Insert the updated hotkeys into the JSON structure and print the resulting JSON
    output_data = {
        "name": keymap_name,
        "version": year_month_day,
        "default_context": "Global Context",
        "os": [
            "windows"
        ],
        "contexts": {"Global Context": updated_hotkeys}}

    with open(output_filename, "w") as out_file:
        json.dump(output_data, out_file, indent=4)

    print(f"JSON data saved to {output_filename}")


if __name__ == "__main__":
    main()

File: sources/game_prefs_to_hotkey_json.js
const fs = require('fs');

function extractHotkeysFromLua(luaContent) {
    // Extract the content between UserKeyMap = { and the closing }
    const userKeyMapPattern = /UserKeyMap\s*=\s*{(.*?)}/s;
    const userKeyMapMatch = luaContent.match(userKeyMapPattern);

    if (!userKeyMapMatch) {
        return [];
    }

    const userKeyMapContent = userKeyMapMatch[1];

    // Extract key-value pairs from userKeyMapContent
    // Ignore the IDE warnings of the regex string, it seems to be correct
    const pattern = /(\[[\'"]?[\w\-\+\s]+[\'"]?\]|[\w\-]+)\s*=\s*'([^\']+)'/g;

    const matches = [...userKeyMapContent.matchAll(pattern)];

    // Convert matches to desired format
    return matches.map(match => [match[1], match[2]]);
}

const luaToJsonFormat = {
    "[": "",
    "]": "",
    "'": "",
    "\"": "",
    "-": " + ",
    "LeftBracket": "[",
    "RightBracket": "]",
    "Equals": "=",
    "NumStar": "*",
    "Quote": "'",
    "NumSlash": "/",
    "Slash": "/",
    "Tilde": "`",
    "Backslash": "\\",
    "Semicolon": ";",
    "Period": "."
};

function convertLuaToJsonFormat(key) {
    for (let luaKeyPart in luaToJsonFormat) {
        while (key.includes(luaKeyPart)) {
            key = key.replace(luaKeyPart, luaToJsonFormat[luaKeyPart]);
        }
    }
    return key;
}

function convertAllKeys(hotkeysArray) {
    return hotkeysArray.map(([key, value]) => {
        const convertedKey = convertLuaToJsonFormat(key);
        return [value, [convertedKey, ""]];
    });
}

// Read the content of the .prefs file
const luaContent = fs.readFileSync('./CheeseBerry_Game.prefs', 'utf8');

// Extract hotkeys
const extractedHotkeys = extractHotkeysFromLua(luaContent);

// Convert hotkeys
const updated_hotkeys = convertAllKeys(extractedHotkeys);

const updatedHotkeysObject = {};
updated_hotkeys.forEach(([key, value]) => {
    updatedHotkeysObject[key] = value;
});

// Print results
//console.log(extractedHotkeys);
//console.log(extractedHotkeys.length);
console.log(updated_hotkeys);

const keymapName = "CheeseBerryJS"; // Replace with your desired keymap name
const yearMonthDay = "2023-10-19"; // Replace with the desired date or dynamically generate it
const outputFilename =`CheeseBerry_${yearMonthDay}_js.json`; // Replace with your desired output filename

const outputData = {
    name: keymapName,
    version: yearMonthDay,
    default_context: "Global Context",
    os: ["windows"],
    contexts: {"Global Context": updatedHotkeysObject}
};

const jsonContent = JSON.stringify(outputData, null, 4);
fs.writeFileSync(outputFilename, jsonContent);
